### PR TITLE
geometry shaders

### PR DESCRIPTION
**Pull request by ColdIce1605**
little broken fov affects it
its going to be marked as draft tell I figure it out

### CODE REVIEW

**Comment by ColdIce1605** [comment]
Bug was do to incorrect inverse matrices fixed in latest iris commits.

File: src/main/java/net/coderbot/iris/gl/program/ProgramBuilder.java [hunk 1]
```diff
@@ -35,15 +36,22 @@ public static ProgramBuilder begin(String name, @Nullable String vertexSource, @
 			throw new RuntimeException("Failed to compile vertex shader for program " + name, e);
 		}
 
+		try {
+			geometry = new GlShader(ShaderType.GEOMETRY, name + ".gsh", geometrySource, EMPTY_CONSTANTS);
+		} catch (RuntimeException e) {
+			throw new RuntimeException("Failed to compile vertex shader for program " + name, e);
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/gl/program/ProgramBuilder.java:42]:
```suggestion
			throw new RuntimeException("Failed to compile geometry shader for program " + name, e);
```

File: src/main/java/net/coderbot/iris/shaderpack/ProgramSource.java [hunk 2]
```diff
@@ -38,7 +44,7 @@ public ProgramSet getParent() {
 	}
 
 	public boolean isValid() {
-		return vertexSource != null && fragmentSource != null;
+		return vertexSource != null && geometrySource != null && fragmentSource != null;
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/shaderpack/ProgramSource.java:47]: This logic is 100% broken. It's entirely valid for a program to lack a geometry shader, and most programs do in fact lack a geometry shader.

File: src/main/java/net/coderbot/iris/gl/program/ProgramBuilder.java [hunk 3]
```diff
@@ -35,15 +36,22 @@ public static ProgramBuilder begin(String name, @Nullable String vertexSource, @
 			throw new RuntimeException("Failed to compile vertex shader for program " + name, e);
 		}
 
+		try {
+			geometry = new GlShader(ShaderType.GEOMETRY, name + ".gsh", geometrySource, EMPTY_CONSTANTS);
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/gl/program/ProgramBuilder.java:40]: This should NOT assume that the geometry shader is always present. Instead, the geometry shader should only be created if it is present, and later should only be attached if it is actually created.

**Review by coderbot16 — CHANGES_REQUESTED** [review]
Geometry shader support is good, however it seems like this code was not tested with shaderpacks that don't use geometry shaders, since I see a few places where things will completely break if a pack does not use geometry shaders.

File: src/main/java/net/coderbot/iris/shaderpack/ProgramSource.java (hunk 2 above)
Comment (ColdIce1605) [reply src/main/java/net/coderbot/iris/shaderpack/ProgramSource.java:47]: what you're saying is it should be removed right?

Comment (coderbot16) [reply src/main/java/net/coderbot/iris/shaderpack/ProgramSource.java:47]: yes, since it is valid for a shader program to lack a geometry shader

**Review by coderbot16 — COMMENTED** [review]
A few stylistic issues but otherwise looks good

File: src/main/java/net/coderbot/iris/gl/shader/ProgramCreator.java [hunk 4]
```diff
@@ -13,7 +13,7 @@ public static int create(String name, GlShader... shaders) {
 		int program = GL20C.glCreateProgram();
 
 		for (GlShader shader : shaders) {
-			GL20C.glAttachShader(program, shader.getHandle());
+			if(shader != null) GL20C.glAttachShader(program, shader.getHandle());
```
[the remark this answers is not in the document]
Comment (ColdIce1605) [reply src/main/java/net/coderbot/iris/gl/shader/ProgramCreator.java:16]: what is the wanted behavior then?

File: src/main/java/net/coderbot/iris/gl/program/ProgramBuilder.java [hunk 5]
```diff
@@ -23,27 +23,35 @@ public void bindAttributeLocation(int index, String name) {
...
 		GlShader vertex;
+		GlShader geometry;
 		GlShader fragment;
 
 		try {
 			vertex = new GlShader(ShaderType.VERTEX, name + ".vsh", vertexSource, EMPTY_CONSTANTS);
 		} catch (RuntimeException e) {
 			throw new RuntimeException("Failed to compile vertex shader for program " + name, e);
 		}
+	if(geometrySource != null)
```
[the remark this answers is not in the document]
Comment (ColdIce1605) [reply src/main/java/net/coderbot/iris/gl/program/ProgramBuilder.java:38]: `
if(geometrySource != null) {
		try {
			geometry = new GlShader(ShaderType.GEOMETRY, name + ".gsh", geometrySource, EMPTY_CONSTANTS);
		} catch (RuntimeException e) {
			throw new RuntimeException("Failed to compile geometry shader for program " + name, e);
		}
	} else {
		geometry = null;
	}
`
is this better?

File: src/main/java/net/coderbot/iris/gl/shader/ProgramCreator.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (coderbot16) [reply src/main/java/net/coderbot/iris/gl/shader/ProgramCreator.java:16]: If a null is passed to ProgramCreator, it should probably crash - nulls shouldn't be passed to ProgramCreator in the first place, imo

File: src/main/java/net/coderbot/iris/gl/program/ProgramBuilder.java [hunk 6]
```diff
@@ -35,15 +36,32 @@ public static ProgramBuilder begin(String name, @Nullable String vertexSource, @
...
+		int programId;
+		
+		if(geometry != null) {
+			programId = ProgramCreator.create(name, vertex, geometry, fragment);
+		} else {
+			programId = ProgramCreator.create(name, vertex, fragment);
+		}
 
 		vertex.destroy();
+		if (geometry != null) { geometry.destroy(); }
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/gl/program/ProgramBuilder.java:64]:
```suggestion

		if (geometry != null) {
			geometry.destroy();
		}

```

File: src/main/java/net/coderbot/iris/gl/program/ProgramBuilder.java [hunk 7]
```diff
@@ -35,15 +36,32 @@ public static ProgramBuilder begin(String name, @Nullable String vertexSource, @
...
 		try {
 			fragment = new GlShader(ShaderType.FRAGMENT, name + ".fsh", fragmentSource, EMPTY_CONSTANTS);
 		} catch (RuntimeException e) {
 			throw new RuntimeException("Failed to compile fragment shader for program " + name, e);
 		}
 
-		int programId = ProgramCreator.create(name, vertex, fragment);
+		int programId;
+		
+		if(geometry != null) {
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/gl/program/ProgramBuilder.java:57]:
```suggestion
		if (geometry != null) {
```

File: src/main/java/net/coderbot/iris/gl/program/ProgramBuilder.java [hunk 8]
```diff
@@ -35,15 +36,32 @@ public static ProgramBuilder begin(String name, @Nullable String vertexSource, @
 			throw new RuntimeException("Failed to compile vertex shader for program " + name, e);
 		}
 
+		if(geometrySource != null) {
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/gl/program/ProgramBuilder.java:39]:
```suggestion
		if (geometrySource != null) {
```

**Review by coderbot16 — CHANGES_REQUESTED** [review]
Looks good aside from a few minor remaining style issues

**Review by coderbot16 — APPROVED** [review]
Thanks!

**Pull request merged**